core/src/main/scala/kafka/log/LocalLog.scala

```diff
@@ -370,11 +370,17 @@ class LocalLog(@volatile private var _dir: File,
         throw new OffsetOutOfRangeException(s"Received request for offset $startOffset for partition $topicPartition, " +
           s"but we only have log segments upto $endOffset.")
 
-      if (startOffset == maxOffsetMetadata.messageOffset)
+      if (startOffset == maxOffsetMetadata.messageOffset) {
         emptyFetchDataInfo(maxOffsetMetadata, includeAbortedTxns)
-      else if (startOffset > maxOffsetMetadata.messageOffset)
+      } else if (startOffset > maxOffsetMetadata.messageOffset ||
+        maxOffsetMetadata.messageOffsetOnly() ||
+        maxOffsetMetadata.segmentBaseOffset < segmentOpt.get.baseOffset) {
```

Comment:
Not sure when do we need this condition (added as per Jun's suggestion):
```
maxOffsetMetadata.segmentBaseOffset < segmentOpt.get.baseOffset
```

Reply:
For the condition `maxOffsetMetadata.segmentBaseOffset < segmentOpt.get().baseOffset()`, perhaps it's better to fold them under the following logic below. This is because we could iterate through multiple segments later.

```
          val maxPosition =
          // Use the max offset position if it is on this segment; otherwise, the segment size is the limit.
            if (maxOffsetMetadata.segmentBaseOffset == segment.baseOffset) maxOffsetMetadata.relativePositionInSegment
            else segment.size

```
We can change the logic to sth like the following.

```
          val maxPosition: Option[Int]  =
            if (segment.baseOffset < maxOffsetMetadata.segmentBaseOffset)
              Some(segment.size)
            else if (segment.baseOffset == maxOffsetMetadata.segmentBaseOffset && !maxOffsetMetadata.messageOffsetOnly()) 
               Some(maxOffsetMetadata.relativePositionInSegment)
            else 
               None
```

We can then change the logic in `segment.read()` to return empty if `maxPosition` is empty.

Reply:
When applying this change, the newly added LocalLog#testWhenFetchOffsetHigherThanMaxOffset test fails for the case-3 and case-4. (ie) Which offset to return back in the FetchDataInfo when the conditions didn't met?

1. In current approach, the same fetch-offset gets returned back in the FetchDataInfo
2. In the suggested approach, the next-offset/log-end-offset gets returned back in the FetchDataInfo

Reply:
Returning the current offset is correct. I am wondering why the suggested approach returns nextOffset. `segment.read(startOffset, maxLength, maxPosition, minOneMessage)` should return a non-null fetchDataInfo since the startOffset exists in the first segment, right?

Reply:
Addressed with the latest commit 77f90f63408ebe5b3da5c3305ad2affb7901eff4